tools/hlsl_generator/gen.py

```diff
@@ -0,0 +1,311 @@
+                case "Barrier" | "Bit":
+                    processInst(writer, instruction, InstOptions())
+                case "Reserved":
+                    match instruction["opname"]:
+                        case "OpBeginInvocationInterlockEXT" | "OpEndInvocationInterlockEXT":
+                            processInst(writer, instruction, InstOptions())
+                case "Non-Uniform":
+                    match instruction["opname"]:
+                        case "OpGroupNonUniformElect" | "OpGroupNonUniformAll" | "OpGroupNonUniformAny" | "OpGroupNonUniformAllEqual":
+                            processInst(writer, instruction, InstOptions(result_ty="bool"))
+                        case "OpGroupNonUniformBallot":
+                            processInst(writer, instruction, InstOptions(result_ty="uint32_t4",op_ty="bool"))
+                        case "OpGroupNonUniformInverseBallot" | "OpGroupNonUniformBallotBitExtract":
+                            processInst(writer, instruction, InstOptions(result_ty="bool",op_ty="uint32_t4"))
+                        case "OpGroupNonUniformBallotBitCount" | "OpGroupNonUniformBallotFindLSB" | "OpGroupNonUniformBallotFindMSB":
+                            processInst(writer, instruction, InstOptions(result_ty="uint32_t",op_ty="uint32_t4"))
+                        case _: processInst(writer, instruction, InstOptions())
+                case _: continue # TODO
+
+        writer.write(foot)
+
+class Shape(Enum):
+    DEFAULT = 0,
+    PTR_TEMPLATE = 1, # TODO: this is a DXC Workaround
+    PSB_RT = 2, # PhysicalStorageBuffer Result Type
+
+class InstOptions(NamedTuple):
+    shape: Shape = Shape.DEFAULT
+    result_ty: Optional[str] = None
+    op_ty: Optional[str] = None
+
+def processInst(writer: io.TextIOWrapper, instruction, options: InstOptions):
+    templates = []
+    caps = []
+    conds = []
+    op_name = instruction["opname"]
+    fn_name = op_name[2].lower() + op_name[3:]
+    result_types = []
+
+    if "capabilities" in instruction and len(instruction["capabilities"]) > 0:
+        for cap in instruction["capabilities"]:
+            if cap == "Shader" or cap == "Kernel": continue
+            caps.append(cap)
+    
+    if options.shape == Shape.PTR_TEMPLATE:
+        templates.append("typename P")
+        conds.append("is_spirv_type_v<P>")
+    
+    # split upper case words
+    matches = [(m.group(1), m.span(1)) for m in re.finditer(r'([A-Z])[A-Z][a-z]', fn_name)]
+
+    for m in matches:
+        match m[0]:
+            case "I":
+                conds.append("(is_signed_v<T> || is_unsigned_v<T>)")
+                break
+            case "U":
+                fn_name = fn_name[0:m[1][0]] + fn_name[m[1][1]:]
+                result_types = ["uint32_t", "uint64_t"]
+                break
+            case "S":
+                fn_name = fn_name[0:m[1][0]] + fn_name[m[1][1]:]
+                result_types = ["int32_t", "int64_t"]
```

Comment:
need a condition about being signed or unsigner

Also result types can be 16 bit ints too!